Create a single top-level docs folder for Servlet and Spring Boot

mraible opened the pull request:
Fixes #949.

To build the docs, cd into `docs` and run `make allhtml`. 

Notes:

1. If I use `.. only:: servlet || springboot` to include certain pages, it results in two table of contents, concatenated together. Servlet doesn’t have `tutorial` or `authorization` pages. It does have `access-control`, which Spring Boot does not.

    ````
    .. only:: servlet

      This part of the documentation will show you how to get started with the Stormpath Java Servlet Plugin.  If you're new to the plugin, start here!

      .. toctree::
        :maxdepth: 2

        about
        quickstart
        config
        registration
        login
        social
        forgot-password
        logout
        http-request-authentication
        request
        access-control
        i18n
        events
        caching
        idsite
        views
        appendix

    .. only:: springboot

      This part of the documentation will show you how to get started right away.  If you are building a Spring Boot application and are new to Stormpath, start here!

      .. toctree::
        :maxdepth: 2

        about
        quickstart
        tutorial
        config
        registration
        login
        social
        authorization
        forgot-password
        logout
        http-request-authentication
        request
        i18n
        events
        caching
        idsite
        views
        appendix
    ````
    
    To fix this, I changed to a single TOC and used `.. only::` at the top of the files. This removes pages from the TOC, but the user is still navigated to these pages with the "Next" button from the previous page. This isn't a great user experience because they see a blank page. The only other solution I can think of is to move these servlet-only or boot-only sections to another page that already exists.

2. Conditional sections in a page causes section to not show up in local TOC, but they do appear in left nav. This means that before the TOC at the top of the QuickStart was:

    ```
    Get an API Key
    Add the Stormpath Default Spring Boot Starter
    Try it!
    Next Steps
    ```
    
    When using a conditional section, this is now rendered as:

    ```
    Get an API Key
    Try it!
    Next Steps
    ```
    
    The good news is it's rendered properly in the left nav.

3. Finally, for some reason, the appendix adds prefixes to a couple items in Spring Boot. Here's the appendix TOC for servlet:

    ```
    A: web.stormpath.properties
    B: i18n.properties
    C: stormpath.css
    ```

    And here's the appendix TOC for Spring Boot:

    ```
    Default Stormpath Properties
    B: i18n.properties
    C: stormpath.css
    /templates/stormpath/change-password.html
    /templates/stormpath/forgot-password.html
    /templates/stormpath/head.html
    /templates/stormpath/login.html
    /templates/stormpath/register.html
    /templates/stormpath/verify.html
    ```

mrioan commented:
@mraible if I execute `make html LANGUAGE=servlet` the Servlet Documentation is properly generated. However it fails when I do `make html LANGUAGE=spring-boot`.

I found out, taking a look at `conf.py` that it works with `make html LANGUAGE=springboot`. So either there is a bug in the configurarion or in the `readme.md` file which instruct us to type `spring-boot`. I am not really sure whether the fix is to simply change the documentation or the `conf.py` file. I am afraid that if I change `conf.py` other thing could be broken. So I leave it up to you to decide which is the real fix.

mrioan commented:
@mraible regarding bullet 1 of the description you added to this PR: I am reading `internationalization` in Servlet plugin. I press the "previous" button at the bottom of the page. I see a blank page but belonging to Spring Boot. I press "previous" button again, then I see `Request Context` for Spring Boot. This means that users will end up reading a different guide (note that I started with Servlet and I ended up reading Spring Boot). Sadly the current TOC approach is not working.

mrioan commented:
@mraible regarding bullet 1 of the description you added to this PR: I am reading `internationalization` in Servlet plugin. I press the "previous" button at the bottom of the page. I see a blank page but belonging to Spring Boot. I press "previous" button again, then I am reading `Request Context` for Spring Boot. This means that users will end up reading a different guide (note that I started with Servlet and I ended up reading Spring Boot). Sadly the current TOC approach is messing things up

I need to clarify that this may be happening because the `make allhtml` command is not working therefore I execute `make html LANGUAGE=servlet` followed by `make html LANGUAGE=springboot` myself. This is causing all the pages to reside in the `build/html` dir (no servlet or springbok dir inside). I think those individual commands should still build the docs in their specific sub-dies rather than in the root of `build/html`. This means that documentations will always reside in the expected location indistinctively of the command used to generate them. This could lead to future issues as new tools/developers will not be aware of this ambiguity.

mrioan commented:
@mraible if I am reading `Request Context` in Servlet and I click the `previous` button (at the bottom) then the `HTTP Request Authentication` is open in a new window (i.e I end up having 2 windows open in my browser)

mrioan commented:
In Spring Boot, if I press the "Default Stormpath Properties" all the properties are expanded in the TOC. Can it be avoided? The experience is not really nice, there are many properties.

mrioan commented:
In Spring Boot, the first option in the Appendix is `Default Stormpath Properties`, Second one is `B: i18n.properties`, third is `C: stormpath.css`, forth is `/templates/stormpath/change-password.html`. I would like to see homogeneity in the appendix. Either not having `B:` and `C:` prefixed, or all of them should have the ordered letter prefixed. I prefer the latter. I think the `Entries` section in http://www.sphinx-doc.org/en/stable/markup/toctree.html solves this

mrioan commented:
@mraible back to you. Thanks!

mraible commented:
@mrioan The `spring-boot` reference in readme.md is a typo. I've fixed that. Not sure why `make allhtml` doesn't work for you. It works fine for me:

```
$ make allhtml
Running Sphinx v1.4.6
making output directory...
loading intersphinx inventory from https://docs.python.org/objects.inv...
intersphinx inventory has moved: https://docs.python.org/objects.inv -> https://docs.python.org/2/objects.inv
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 32 source files that are out of date
updating environment: 32 added, 0 changed, 0 removed
reading sources... [100%] views
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] views
generating indices... genindex
writing additional pages... search
copying images... [100%] _static/linkedin-add-authorized-urls.gif
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.
Running Sphinx v1.4.6
making output directory...
loading intersphinx inventory from https://docs.python.org/objects.inv...
intersphinx inventory has moved: https://docs.python.org/objects.inv -> https://docs.python.org/2/objects.inv
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 32 source files that are out of date
updating environment: 32 added, 0 changed, 0 removed
reading sources... [100%] views
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] views
generating indices... genindex
writing additional pages... search
copying images... [100%] _static/linkedin-add-authorized-urls.gif
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.
Build finished. The HTML pages are in build/html.
```

mrioan commented:
I see `Sphinx v1.4.1` for me while `1.4.6` for you. It seems my version is automatically treating warnings as errors while yours either is not finding that same warning or it does not treat them as errors:
```
Warning, treated as error:
/Users/mario/projects/stormpath/iam/stormpath-sdk-java/docs/source/appendix/spring-boot-core-properties.rst:: WARNING: document isn't included in any toctree
```
I tried upgrading but had dependencies issue. If it works for you then I am ok, this seems to be just an 
issue in my machine

mrioan commented on docs/readme.md:
```diff
@@ -10,6 +10,8 @@ If you already have Sphinx, then just upgrade:
 
 `pip install -U Sphinx sphinx_rtd_theme`
 
+**NOTE:** Make sure you are using Sphinx 1.4.6 as a minimum. With previous versions, warnings are treated as errors.
```
Awesome, thanks!

jakubswiatczak commented:
So first of all, in commit 0a26ecb3e75a1ef1117f685dcef8d2cf2a96c636 I removed the "Next" and "Previous" buttons entirely. I did this by adding this to `confy.py`:

```
html_context = {
    'css_files': [
        '_static/theme_overrides.css',  # overrides for wide tables in RTD theme
        ],
}
```

and then in `/_static/theme_overrides.css` I added this:

```
.rst-footer-buttons {
    display:none;
}
```

mrioan commented:
I discussed [this issue](https://github.com/stormpath/stormpath-sdk-java/pull/958#issuecomment-245912558) with @jakubswiatczak and we think that there are a two options:

1. Group properties in some categories. This way, the first-level header will have much fewer items
2. These properties could be automatically generated if the metada json files are converted to html and imported during docs generation

Bullet 1 still seems very difficult to maintain and will eventually lose sync with real properties therefore we both think that the way to go is bullet 2

mraible commented:
@mrioan The properties documentation is generated by a test, so we can easily manipulate the output. You can see the test at https://github.com/stormpath/stormpath-sdk-java/blob/949-consolidate-docs/extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/autoconfigure/StormpathAutoConfigurationVersusMetadata.groovy.

mraible commented:
@mrioan I removed the TOC from the properties page and removed "B" and "C" from the titles of the i18n and CSS pages (in the appendix as well). Please review again when you get a chance.

mrioan commented:
I am documenting here the solution to the following error when building docs:
```
Running Sphinx v1.4.6
making output directory...

Theme error:
no theme named 'stormpath' found (missing theme.conf?)
Running Sphinx v1.4.6
making output directory...

Theme error:
no theme named 'stormpath' found (missing theme.conf?)
make: *** [allhtml] Error 1
```

Solution: Run the following command: `git submodule update --init --recursive`. That's it; you can build docs now.

mrioan commented:
@mraible back to you. Thanks

mraible commented:
@mrioan I've fixed both of the issues you mention.

mrioan commented:
@mraible It all looks just fine to me. I have a pending question though.

WRT to [this answer](https://github.com/stormpath/stormpath-sdk-java/pull/958#issuecomment-249614130). Does it mean that the documentation for these properties are already automatically generated? Or we are still needing some more step for them to be totally automated? Thanks

mraible commented:
@mrioan Yes, the documentation for these properties is automatically generated by running two tests. 

* core properties: [StormpathAutoConfigurationVersusMetadata.groovy](https://github.com/stormpath/stormpath-sdk-java/blob/949-consolidate-docs/extensions/spring/boot/stormpath-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/autoconfigure/StormpathAutoConfigurationVersusMetadata.groovy)
* web properties: [StormpathWebAutoConfigurationVersusMetadata.groovy](https://github.com/stormpath/stormpath-sdk-java/blob/949-consolidate-docs/extensions/spring/boot/stormpath-webmvc-spring-boot-starter/src/test/groovy/com/stormpath/spring/boot/autoconfigure/StormpathWebAutoConfigurationVersusMetadata.groovy)

mrioan commented:
@mraible the javadoc for those tests say that "they **compare** blah". But they do not say that they **automatically generate** documentation. 

Can you please add that to the javadoc as part of this PR? If those tests impact documentation that users will read we need to be very careful with them. Please also add a link to issue 949. 

Once you do that you can merge. Thanks @mraible
